### PR TITLE
Shim OpenSSL Evp_Cipher* methods.

### PR DESCRIPTION
**Pull request by eerhardt**
Adding shims for all OpenSSL methods in Interop.EVP.Cipher.cs.

@bartonjs, @nguerrera, @sokket

### CODE REVIEW

File: src/Common/src/Microsoft/Win32/SafeHandles/SafeEvpCipherCtxHandle.Unix.cs [hunk 1]
```diff
@@ -37,11 +37,11 @@ internal static SafeEvpCipherCtxHandle Create()
 
             try
             {
-                memPtr = Marshal.AllocHGlobal(Interop.libcrypto.EVP_CIPHER_CTX_SIZE);
+                memPtr = Marshal.AllocHGlobal(Interop.Crypto.EVP_CIPHER_CTX_SIZE);
```
Comment (bartonjs) [inline src/Common/src/Microsoft/Win32/SafeHandles/SafeEvpCipherCtxHandle.Unix.cs]: Actually, while you're in here; could you take care of #1963, and make Create and Free methods for this type?  Then we can get rid of EVP_CIPHER_CTX_SIZE.

Comment (bartonjs) [inline src/Common/src/Microsoft/Win32/SafeHandles/SafeEvpCipherCtxHandle.Unix.cs]: I guess you sort of did already, by making GetEvpCipherCtxSize.  I think I'm okay with that...

Should still tag the issue as fixed, though :)

**Comment by nguerrera** [comment]
Rather than XxxGetSize(), AllocHGlobal(), XxxInit(), XxxFinal(), FreeHGlobal() sequence, I think there's value in a slight 1:1 departure with XxxCreate(), XxxDestroy(). That was what I found nicer for HmacXxx(). I forget if I missed that same pattern on last review, but it would be good to be consistent whenever OpenSSL has a similar pattern.

File: src/Common/src/Microsoft/Win32/SafeHandles/SafeEvpCipherCtxHandle.Unix.cs (hunk 1 above)
Comment (nguerrera) [inline src/Common/src/Microsoft/Win32/SafeHandles/SafeEvpCipherCtxHandle.Unix.cs]: I personally like making less chatty Create/Destroy shims like what I did for HMAC, but if you guys think 1:1 with GetSize and Alloc/FreeHGlobal is better, I won't push back, but then you should probably switch my HMAC code to match this pattern...

**Comment by nguerrera** [comment]
LGTM apart from the create/destroy vs 1:1 + get size/allochglobal/freehglobal discussion. I'll leave it to @bartonjs and your judgment, which patter you prefer.

**Comment by stephentoub** [comment]
@dotnet-bot test this please (https://github.com/dotnet/corefx/issues/3682)

File: src/Common/src/Microsoft/Win32/SafeHandles/SafeEvpCipherCtxHandle.Unix.cs (hunk 1 above)
Comment (bartonjs) [inline src/Common/src/Microsoft/Win32/SafeHandles/SafeEvpCipherCtxHandle.Unix.cs]: I think I'd prefer a Create and Free/Destroy/whatever within the PAL.  In particular, if OpenSSL got around to adding methods later we'd (maybe) want to opportunistically just shim it.

**Comment by eerhardt** [comment]
@nguerrera, @bartonjs - the EVP_MD_CTX* uses a very slightly different pattern.

Example:
```C++
EVP_MD_CTX *mdctx = EVP_MD_CTX_create();
EVP_DigestInit_ex(mdctx, md, NULL);
EVP_DigestUpdate(mdctx, mess1, strlen(mess1));
EVP_DigestUpdate(mdctx, mess2, strlen(mess2));
EVP_DigestFinal_ex(mdctx, md_value, &md_len);
EVP_MD_CTX_destroy(mdctx);
```

Here, they have Create() & Destroy() methods, and you need to call the Init_ex method after Create().  So I could combine Create() and Init_ex() into a single P/Invoke.  Do you think that is beneficial?  It would be one less P/Invoke.  I could then turn our Init_ex() method into a "Reset()", similar to HMAC and what I did for Interop.EVP.Cipher.

If you want, I can add that change to this PR.  Let me know.

**Comment by bartonjs** [comment]
LGTM :)

**Comment by stephentoub** [comment]
LGTM2

**Pull request merged**